File: TripKitUI/map managers/TripMapManager.m
//
//  TripMapManager.m
//  Pods
//
//  Created by Adrian Schoenig on 7/09/2016.
//
//

#import "TripMapManager.h"

#ifdef TK_NO_FRAMEWORKS
#import <TripKit/TKTripKit.h>
#import <TripKit/TripKit-Swift.h>
#else
#import <TripKitUI/TripKitUI-Swift.h>
#endif

#define TIME_BETWEEN_UPDATES 10

@interface TripMapManager ()

@property (nonatomic, strong) TKBuzzInfoProvider *infoProvider;

@end

@implementation TripMapManager

#pragma mark - RouteMapManager

- (instancetype)init
{
  self = [super init];
  if (self) {
    self.alerts = @[];
  }
  return self;
}

- (void)takeChargeOfMap:(MKMapView *)mapView
      forViewController:(UIViewController<ASMapManagerDelegate> *)controller
               animated:(BOOL)animated
             completion:(void (^)(BOOL finished))completion
{
  [super takeChargeOfMap:mapView forViewController:controller animated:animated completion:nil];
  
  // We always want to zoom at first
  self.forceZoom = YES;
  
  if (self.trip) {
    [self showRoute:self.trip animated:animated];
  }
  
  if (completion != nil) {
    completion(YES);
  }
}

- (BOOL)drawAsLargeCircle:(id<MKAnnotation>)annotation
{
#pragma unused(annotation)
  return NO; // we already show segment pins
}

- (BOOL)drawCircleAsTravelled:(id<MKAnnotation>)annotation
{
  if ([annotation isKindOfClass:[StopVisits class]]) {
    return [self.trip usesVisit:(StopVisits *)annotation];
  } else {
    return NO;
  }
}

- (BOOL)zoomToAnnotation:(id<MKAnnotation>)annotation
{
  if ([annotation isKindOfClass:[StopVisits class]]) {
    return [self.trip usesVisit:(StopVisits *)annotation];
  } else {
    return YES;
  }
}

#pragma mark - Showing the trip

- (void)reloadCurrentTrip
{
  [self refreshRouteWithAnimated:NO forceRebuild:YES zoom:ZoomModeZoomOnly];
}

- (void)showSegment:(TKSegment *)segment
           animated:(BOOL)animated
{
  ZAssert(nil != segment, @"TripMapManager needs a segment to display!");
  
  if (self.trip != segment.trip) {
    [self showRoute:segment.trip animated:animated];
    return;
  }
  
  [self refreshRouteWithAnimated:NO forceRebuild:NO zoom:ZoomModeZoomOnly];
  
  [self deselectAll];
  [self.mapView selectAnnotation:segment animated:YES];
  
  NSArray *annotationsToZoomTo = [segment annotationsToZoomToOnMap];
  if (annotationsToZoomTo) {
    [self.mapView zoomToAnnotations:annotationsToZoomTo
                        edgePadding:[self.delegate mapManagerEdgePadding:self]
                           animated:animated];
  }
}

- (void)showRoute:(Trip *)newTrip animated:(BOOL)animated
{
  [self show:newTrip zoom:ZoomModeAddAndZoomToTrip animated:animated];
}

#pragma mark - Alerts

- (void)selectAlert:(Alert *)alert
{
  if (alert.location) {
    [self.mapView selectAnnotation:alert animated:YES];
  } else {
    TKSegment *segment = [self segmentForAlert:alert];
    [self.mapView selectAnnotation:segment animated:YES];
  }
}

- (TKSegment *)segmentForAlert:(Alert *)alert
{
  for (TKSegment *segment in self.trip.segments) {
    for (Alert *ale in segment.alerts){
      if ([alert.hashCode isEqualToNumber:ale.hashCode]) {
        return segment;
      }
    }
  }
  ZAssert(false, @"Alert matches no segment: %@", alert);
  return nil;
}



#pragma mark - RealTime

- (void)realTimeUpdateForTrip:(Trip *)theTrip animated:(BOOL)animated
{
  NSMutableArray *newAlerts = [NSMutableArray array];
  
  NSMutableArray<Vehicle *> *primaries = [NSMutableArray array];
  NSMutableArray<Vehicle *> *secondaries = [NSMutableArray array];
  
  for (TKSegment *segment in theTrip.segments) {
    segment.time = segment.time; // KVO :(
    segment.title = @""; // KVO :(
    
    Vehicle *primary = segment.realTimeVehicle;
    if (primary) {
      [primaries addObject:primary];
    }
    NSArray *secondary = segment.realTimeAlternativeVehicles;
    if (secondary) {
      [secondaries addObjectsFromArray:secondary];
    }
    
    for (Alert *alert in segment.alertsWithLocation) {
      if (! [self.alerts containsObject:alert]) {
        [newAlerts addObject:alert];
      }
    }
  }
  
  [self realTimeUpdateForPrimaryVehicles:primaries
                       secondaryVehicles:secondaries
                                animated:animated];
  
  [UIView animateWithDuration:animated ? 1 : 0
                   animations:
   ^{
     [self addAnnotations:newAlerts];
     [self.mapView addAnnotations:newAlerts];
   }];
}


#pragma mark - Lazy

- (TKBuzzInfoProvider *)infoProvider {
  if (!_infoProvider) {
    _infoProvider = [[TKBuzzInfoProvider alloc] init];
  }
  return _infoProvider;
}


@end

File: TripKitUI/map managers/TripMapManager.swift
//
//  RouteMapManager+Server.swift
//  TripKit
//
//  Created by Adrian Schoenig on 7/09/2016.
//
//

import Foundation
import MapKit

#if TK_NO_MODULE
#else
  import TripKit
#endif


@objc
public enum ZoomMode: Int {
  case zoomOnly
  case addAndZoomToStart
  case addAndZoomToTrip
}

extension TripMapManager {

  @objc
  public func show(_ trip: Trip, zoom: ZoomMode, animated: Bool = false) {
    guard let mapView = mapView() else { return }
    
    // Make sure we don't show that trip already
    guard trip != self.trip || mapView.annotations.isEmpty else { return }
    
    deselectAll()
    self.trip = trip
    
    var forzeZoom = self.forceZoom
    if !forzeZoom
      && mapView.visibleMapRect.origin.x >= 0 // might happen if map view is minimized
    {
      let fromToRect = [trip.request.fromLocation, trip.request.toLocation].mapRect
      forzeZoom = (!MKMapRectIntersectsRect(mapView.visibleMapRect, fromToRect))
    }
    
    // don't force zooma gain
    self.forceZoom = false
    
    let zoomMode = self.forceZoom ? .addAndZoomToTrip : zoom
    refreshRoute(animated: animated, forceRebuild: true, zoom: zoomMode)
  }
  
  @objc
  public func refreshRoute(animated: Bool, forceRebuild force: Bool, zoom: ZoomMode) {
    guard let mapView = mapView(), let trip = self.trip else { return }
    
    if UIDevice.current.userInterfaceIdiom == .phone {
      assert(mapView.delegate === self, "We are not the map view's delegate, but \(String(describing: mapView.delegate)) is.")
    }
    
    if force {
      removeLastRouteFromMap()
      if self.trip == nil {
        return
      }
    }
    
    add(trip, zoom: zoom, animated: animated)
  }
  
  private func add(_ trip: Trip, zoom: ZoomMode = .addAndZoomToTrip, animated: Bool = false) {
    guard let mapView = mapView() else { return }
    
    var zoomTo = [MKAnnotation]()
    var primaries = [Vehicle]()
    var alternatives = [Vehicle]()
    var affectedByTraffic = false
    var alerts = [Alert]()
    
    var addedSegments = 0
    for segment in trip.segments() {
      if (segment as STKDisplayablePoint).pointDisplaysImage {
        if zoom != .zoomOnly {
          add(segment)
        }
        
        // zoom to this segment if we're not in 'zoom to start'
        // mode OR if we're in 'zoom to start' mode and this is
        // near the start
        if zoom != .addAndZoomToStart
          || (segment.order() == .regular && addedSegments < 2)
          || (segment.order() == .end && addedSegments == 1) {
          zoomTo.append(segment)
          addedSegments += 1
        }
        
        guard zoom != .zoomOnly, let toAdd = MapManagerHelper.shapeAnnotations(for: segment) else { continue }

        add(toAdd.points)
        add(toAdd.overlays)
        if toAdd.requestVisits {
          self.requestVisits(for: segment, includeShape: toAdd.overlays.isEmpty)
        }
        
        // vehicles
        if let primary = segment.realTimeVehicle() {
          primaries.append(primary)
        }
        alternatives += segment.realTimeAlternativeVehicles()
        
        // alerts
        let segmentAlerts = segment.alertsWithLocation()
        add(segmentAlerts)
        alerts.append(contentsOf: segmentAlerts)
        
        affectedByTraffic = affectedByTraffic || segment.isAffectedByTraffic()
      }
    }
    
    zoomTo += primaries.map { $0 as MKAnnotation }
    addPrimaryVehicles(primaries, secondaryVehicles: alternatives)
    
    if #available(iOS 9.0, *) {
      mapView.showsTraffic = affectedByTraffic
    }
    
    if zoomTo.count > 0 {
      mapView.zoom(to: zoomTo, edgePadding: delegate()?.mapManagerEdgePadding(self) ?? UIEdgeInsets.zero, animated: animated)
    }
    
    self.alerts = alerts
    
    if zoom != .zoomOnly {
      presentRoute()
    }
    
    // Note: this does not select, but the caller might want to select the main segment
  }
  
  private func requestVisits(for segment: TKSegment, includeShape: Bool) {
    guard
      segment.isPublicTransport(),
      let service = segment.service(),
      let region = segment.startRegion(),
      let departure = segment.departureTime
      else { return }
    
    infoProvider.downloadContent(of: service, forEmbarkationDate: departure, in: region) { [weak self] (updatedService, finished) in
      
      guard let `self` = self else { return }
      guard service == updatedService && finished else { return }
      
      if includeShape, let startCode = segment.scheduledStartStopCode(), let endCode = segment.scheduledEndStopCode() {
        
        
        let start = service.visit(forStopCode: startCode)
        let end = service.visit(forStopCode: endCode)
        for shape in service.shapes(forEmbarkation: start, disembarkingAt: end) {
          self.addShape(shape)
        }
      }
      
      for visit in service.visits ?? [] {
        if segment.shouldShowVisit(visit) {
          self.addAndPresent(visit)
        }
      }
      
    }
  }
  
}
